pkg/cmd/root/completion.go

```diff
@@ -0,0 +1,236 @@
+
+func completionWrapper(completer cobraCompletionFuncWithError) cobraCompletionFunc {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		// By default, the factory will provide a target manager that uses the DynamicTargetProvider (DTP)
+		// implementation, i.e. is based on the file just as much as the CLI flags.
+		// The DTP tries to allow users to "move up", i.e. when they already targeted a shoot, just adding
+		// "--garden foo" should not just change the used garden cluster, but _target_ the garden (instead
+		// of the shoot). This behaviour is not suitable for the CLI completion functions, because
+		// when completing "gardenctl --garden foo --shoot [tab]", the DTP would consider this as
+		// "user wants to target the garden" and will therefore throw away the project/seed information.
+		// Project and seed information however are important for the completion functions.
+		//
+		// To work around this, all completion functions use the regular filesystem based target provider.
+		factory.TargetFile = targetProvider.TargetFile
+		factory.TargetProvider = nil
+
+		result, err := completer(cmd, args, toComplete)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "%v\n", err)
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+
+		return util.FilterStringsByPrefix(toComplete, result), cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+func gardenFlagCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, error) {
+	cfg, err := config.LoadFromFile(factory.ConfigFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load configuration: %w", err)
+	}
+
+	names := sets.NewString()
+	for _, garden := range cfg.Gardens {
+		names.Insert(garden.Name)
+	}
+
+	return names.List(), nil
+}
+
+func projectFlagCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, error) {
+	manager, err := factory.Manager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create manager: %w", err)
+	}
+
+	// any --garden flag has precedence over the config file
+	var currentTarget target.Target
+
+	if targetProvider.GardenNameFlag != "" {
+		currentTarget = target.NewTarget(targetProvider.GardenNameFlag, "", "", "")
+	} else {
+		currentTarget, err = manager.CurrentTarget()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read current target: %w", err)
+		}
+	}
+
+	gardenClient, err := manager.GardenClient(currentTarget)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client for garden cluster %q: %w", currentTarget.GardenName(), err)
+	}
+
+	projectList := &gardencorev1beta1.ProjectList{}
+	if err := gardenClient.List(factory.Context(), projectList); err != nil {
+		return nil, fmt.Errorf("failed to list projects on garden cluster %q: %w", currentTarget.GardenName(), err)
+	}
+
+	names := sets.NewString()
+	for _, project := range projectList.Items {
+		names.Insert(project.Name)
+	}
+
+	return names.List(), nil
```

Comment:
once the current target is determined, this code looks the same as for the argument completion. Same for the other completions. Doesn't it make sense to de-duplicate even more code?

Reply:
I don't quite understand what you code blocks you mean.

Reply:
https://github.com/gardener/gardenctl-v2/pull/21/files/240f1aeb8acd2c5ee9f78cb9966692d53940b9c0#diff-517d5acc56e30d22994a8901475305ae60213bb2598de53d75c92da4e14d4bc8R145-R160
```go
	gardenClient, err := manager.GardenClient(currentTarget)
	if err != nil {
		return nil, fmt.Errorf("failed to create Kubernetes client for garden cluster %q: %w", currentTarget.GardenName(), err)
	}

	projectList := &gardencorev1beta1.ProjectList{}
	if err := gardenClient.List(factory.Context(), projectList); err != nil {
		return nil, fmt.Errorf("failed to list projects on garden cluster %q: %w", currentTarget.GardenName(), err)
	}

	names := sets.NewString()
	for _, project := range projectList.Items {
		names.Insert(project.Name)
	}

	return names.List(), nil
```

https://github.com/gardener/gardenctl-v2/pull/21/files/240f1aeb8acd2c5ee9f78cb9966692d53940b9c0#diff-ace14d9438fa14f933287b67235812a949d856032fd22ea809868e74679f1ccdR76-R91
```go
	gardenClient, err := manager.GardenClient(t)
	if err != nil {
		return nil, fmt.Errorf("failed to create Kubernetes client for garden cluster %q: %w", t.GardenName(), err)
	}

	projectList := &gardencorev1beta1.ProjectList{}
	if err := gardenClient.List(ctx, projectList); err != nil {
		return nil, fmt.Errorf("failed to list projects on garden cluster %q: %w", t.GardenName(), err)
	}

	names := sets.NewString()
	for _, project := range projectList.Items {
		names.Insert(project.Name)
	}

	return names, nil
```